pymobiledevice3/services/afc.py

```diff
@@ -208,36 +208,43 @@ def __init__(self, lockdown: LockdownClient, service_name='com.apple.afc'):
         super().__init__(lockdown, service_name)
         self.packet_num = 0
 
-    def pull(self, src: str, dst: str, callback=None):
+    def pull(self, src: str, dst: str, callback: Optional[callable] = None):
         """ pull a file/directory from given path into a local file/directory """
-        if callback is not None:
-            callback(src, dst)
-
         src = self.resolve_path(src)
-
         if not self.isdir(src):
             # normal file
-            if os.path.isdir(dst):
-                dst = os.path.join(dst, os.path.basename(src))
-            with open(dst, 'wb') as f:
-                f.write(self.get_file_contents(src))
+            if dst.endswith(posixpath.sep):
+                dst_parent_dir = os.path.realpath(os.path.dirname(dst))
```

Comment:
```
dst_path = pathlib.Path(dst).expanduser().resolve()
if dst.endswith(os.sep):
    dst_path = dst_path / os.path.basename(src)
if not dst_path.parent.is_dir():  # check if the parent dir exist
    raise NotADirectoryError(f'pull: directory {dst} does not exist')
dst_path.write_bytes(self.get_file_contents(src))
```

Reply:
this block have couple of bugs, like pull src file to existing folder without trailing slash.

but i took inspiration from this Implementation:
```python
if dst.endswith((os.path.sep, posixpath.sep)) and not dst_path.is_dir():
    raise NotADirectoryError(f'pull: directory {dst_path} does not exist')

if dst_path.is_dir():
    dst_path = dst_path.joinpath(os.path.basename(src))

if not dst_path.parent.is_dir():
    raise FileNotFoundError(f'pull: {dst_path}: No such file or directory')
```

also, in this Implementation you distinguish between the errors :)

Reply:
the last part (`if not dst_path.parent.is_dir():`) move to be in the start of the function :)